README.markdown
# HOLLy - Hypothesised Object from Light Localisations
## A neural network for 3D Reconstruction from 2D Microscopy images

[![DOI](https://zenodo.org/badge/369292185.svg)](https://zenodo.org/badge/latestdoi/369292185)

![Evolving structure](https://shutr.benjamin.computer/static/shutr_files/original/num2021_05_14_cep_0.gif)
![Bunny](https://shutr.benjamin.computer/static/shutr_files/original/file.gif)

A [PyTorch](https://pytorch.org/) based neural network program designed to recontruct protein complexes in 3D from a series of 2D microscopy images. We predict the pose and structure using a Convolutional Neural Network. 

## Overview

HOLLy takes a number of images (it can also generate images from a ground-truth point-cloud) and trains on these attempting to improve it's own internal representation of what it thinks the ground-truth object is. At the end of training, an *obj* or **ply* file will be produced, representing the object the network converged upon.

HOLLy is split into *train.py* and *run.py* with the actual net stored in in *net/net.py*. *eval.py* will evaluate a trained network, creating statistics and visualistions. 

*net/renderer.py* contains the code for the differentiable renderer. *data/loader.py*, along with *data/buffer.py* and *data/batcher.py* create our simulated data for all the tests, including adding noise. *data/imageload.py* is similar, but for the real microscopy images.

(The diagram below is animated and may take a little time to appear).
![Overview diagram](https://shutr.benjamin.computer/static/shutr_files/original/diagram.gif)

## Documentation

This readme covers most of the details of HOLLy, but we also have more documentation available at [Read The Docs](https://holly.readthedocs.io/en/latest/).

## Installation

### Requirements

Requirements include:

* A Linux system setup to perform deeplearning, with appropriate nvidia drivers installed.
* Python version 3
* CUDA setup and running
* A GPU with at least 4G of memory (more if you are fitting more points).

And one of the following

* [Miniconda](https://docs.conda.io/en/latest/miniconda.html).
* [pyenv/virtualenv](https://github.com/pyenv/pyenv).
* [Docker](https://www.docker.com/).

If you want to generate lots of statistics for use later you'll also need all of the following installed and running. Initially, you shouldn't need these but you might see a few warnings pop up:

* [PostgreSQL](https://www.postgresql.org/)
* [Redis](https://redis.io/)

If you want some fancy formatted results or if you are running the tl;dr script, install the following:

* [Imagemagick](https://imagemagick.org/index.php)
* [ffmpeg](https://ffmpeg.org/)

Finally, in order to view the model the network has derived, you'll need a program capable of displaying obj or ply files. These are pretty numerous, but here are a few we use:

* [meshlab](https://www.meshlab.net/)
* [blender(https://www.blender.org/)

Chances are, if you are running a Linux distribution, you will have these already and if not, they'll be in your repository management system (like apt or pacman).

All of the python requirements are listed in the requirements.txt file (there aren't too many).

### tl;dr version

Run the script *tldr.sh* from the top directory. You will need [miniconda](https://docs.conda.io/en/latest/miniconda.html) installed and a computer running Linux, capable of trainin a large neural network with Pytorch. As a guide, a machine with several cores, >=16G memory and an nVida GPU such as a GTX 2080Ti will work reasonably well.

The script creates a miniconda environment called "holly", downloads packages and starts running an experiment, followed by a generation of results and a *final.ply* file representing the learned model. Once this script has completed, take a look at the *Outputs* section of this readme to understand what the network has produced.

### via miniconda

Assuming you have [installed miniconda](https://docs.conda.io/en/latest/miniconda.html), start with the following:

    conda create -n holly python
    conda activate holly
    pip install -r requirements.txt

### via docker

There is a dockerfile included in *docker/holly* that will create a container that you can use to train our network.

Assuming you have docker installed, you can create the docker container like so:

    docker build -t holly docker/holly

## Training

### Simulated data

We can simulate data from a [wavefront obj file](https://en.wikipedia.org/wiki/Wavefront_.obj_file), several of which are included with this repository. Alternatively, you can provide your own. 

To train a new network on simulated data, run the following command in the top level of the project, assuming you are inside the miniconda environment:

    python train.py --obj objs/teapot_large.obj --train-size 40000 --lr 0.0004 --savedir /tmp/test_run --num-points 230 --epochs 20 --sigma-file run/sigma_quick.csv

It is also possible to use the provided bash script to execute training in a given directory, saving the various options and code versions to various files. This is more useful if you want to run many tests, saving the various settings.

    mkdir ../runs/temp
    cd run
    cp run.conf.example ../../runs/temp/run.conf
    ./train.sh ../../runs/temp "My first run"

The bash script *train.sh* looks for a file called *run.conf* that contains the required data for training a network. The file *run.conf.example* can be copied to *run.conf*. Make sure the directory referred to for saving the results exists.

Using docker, one can run the same command as follows:

    mkdir experiment
    docker run --gpus=all --volume="$PWD:/app" holly python train.py --obj objs/teapot_large.obj --train-size 40000 --lr 0.0004 --savedir /tmp/experiment --num-points 230 --epochs 20 --sigma-file run/sigma_quick.csv

Confirm that docker can see the gpu:
    docker run --gpus all holly nvidia-smi

Your GPU should be listed.

Docker needs the [nvidia-docker2](https://github.com/NVIDIA/nvidia-docker) package in order to access gpus. On Ubuntu 20.04 LTS one needs to run the following commands:

    curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | apt-key add -
    curl -s -L https://nvidia.github.io/nvidia-docker/ubuntu20.04/nvidia-docker.list > /etc/apt/sources.list.d/nvidia-docker.list
    apt update
    apt -y install nvidia-docker2
    systemctl restart docker 

There are a large number of options you can pass to the network. These are listed in the *train.py* file. The most important are listed below:

* --save-stats - save on-going train and test statistics.
* --obj - when simulating data, this is the path to the object file we want to use.
* --train-size - the size of the training set
* --lr - the learning rate
* --num-points - how many points do we want to try and fit to the structure.
* --epochs - how many epochs should we run for?
* --sigma-file - the path to the file containing the per-epoch sigmas.
* --image-width - the width of the input and output images
* --image-height - the height of the input and output images

Training usually takes around 4 hours on a nVidia 2080Ti system when running for 20 epochs on an 80,000 size training set.

The final structure discerned by the network is saved as both an obj and a ply file. Both can be viewed in a program like [Meshlab](https://www.meshlab.net/) or [Blender](https://www.blender.org/). The files can be found in:

    <save_directory>/plys/shape_e<epoch>_s<step>.ply
    <save_directory>/objs/shape_e<epoch>_s<step>.obj

### Real Image data

To use real, experimental data one must have a number of images, rendered using a particular sigma. These images need to be held within a directory stucture that matches the sigma file passed in. For example:

    images
    |___10.0
        |____image0.fits
        |____image1.fits
    |___8
        |____image0.fits
        |____image1.fits
    |___1.2
        |____image0.fits
        |____image1.fits

The sigma file (sigma_images.csv) would be:
    10,8,1.2

Images should be in [FITS Format](https://en.wikipedia.org/wiki/FITS) (they support floating point values).
We would then run the following command:
    python train.py --fitspath images --train-size 80000 --lr 0.0004 --savedir /tmp/runs/test_run --num-points 230 --no-translation --epochs 20 --sigma-file sigma_images.csv

A test data set is available for download from [Zenodo](https://zenodo.org/record/4751057). This is a big download - just under 50G! However it does have all the data ready for use, pre-rendered.  

Once you've donwloaded and unzipped the data, place all the images into a directory structure like this:

    holly
    |___run
           |___paper
                    |____10.0
                    |____8
                    etc
                    ...

Use the script *run_cp.sh* in the *run* directory. This assumes you've installed the environment with either miniconda or docker. This script assumes you have downloaded the dataset and placed it the correct directory.

If you want to see the results of a real CEP152 run - the ones from our paper - you can download them from [Zenodo](https://zenodo.org/record/4836173) too!

#### Generating your own images from STORM localisations.

The data used in the paper comes from [Suliana Manley's research group](https://www.epfl.ch/labs/leb/) and is available as a MATLAB/HDF5 file. In order to use it with this network, you will need to render the data to a series of FITs images. We have written a program to do this called [CEPrender](https://github.com/OniDaito/CEPrender). You can download and install this program, then generate the data as follows:

    cargo run --release --bin render /data/Cep152_all.mat /paper/10 24 10 10 cep_all_accepted.txt
    cargo run --release --bin render /data/Cep152_all.mat /paper/9 24 9 10 cep_all_accepted.txt
    ...
    ..
    .
    cargo run --release --bin render /data/Cep152_all.mat /paper/1.41 24 1.41 10 cep_all_accepted.txt

... filling the missing steps with the other sigma levels. This takes quite a while, even with only 10 augmentations. The pre-rendered data can be found on [Zenodo](https://zenodo.org/record/4751057) complete the with the full instructions for generating it.

### Energy Usage

The estimated energy use to train a simulated model is 2.1kWh based on a measurement of 623.4kWh over 166 days. In this period, 298 models were trained and evaluated. This was confirmed by cross-checking against the wattage of the GPU and the time spent to generate an average model.

## Outputs

Once you have a trained network, you can generate the final outputs using the *generate_stats.sh* script found in the *run* directory.

    ./generate_stats.sh  -i <path to your results directory> -g <path to object file> -a -z

So with the test example given above:
    
    ./generate_stats.sh  -i /tmp/runs/test_run -g ../objs/teapot.obj -a -z

This relies on having imagemagick 'montage' command and the correct python libs installed.

This script loads the network and tests it against the ground_truth (if we are using simulated data)

## Training and Test statistics

Due to the nature of this work, we used [Jupyter Notebooks](https://jupyter.org/), Redis and PostgreSQL to save our various statistics, using the code found in *stats/stats.py*. By default, these statistics include:

* Train and test errors.
* Input translations and rotations that make up the dataset.
* Output translations and rotations the network decided upon.
* Input and output sigma.

The training machine must be running an instance of either Redis or PostgreSQL. If both are running, the program will use both. Imagemagick and ffmpeg are used to create montage images and animations.

At some point in the future, we might move to [Tensorboard](https://www.tensorflow.org/tensorboard/) and / or [Weights and Biases](wandb.ai/).

## Tests

Our code comes with a considerable number of tests. These can be run from the top level directory as follows:

    python test.py

Individual tests can be run as follows:

    python -m unittest test.data.Data.test_wobble

## Datasets in the Publication

The CEP152 data we used in the paper with HOLLy can be found on Zenodo at [https://zenodo.org/record/4751057](https://zenodo.org/record/4751057) and the results we mention in the paper can also be downloaded from Zenodo at [https://zenodo.org/record/4836173](https://zenodo.org/record/4836173).

## Command line and configuration options

When running train.py, there are a number of options one can choose.

    --batch-size
    Input batch size for training (default: 20).

    --epochs
    Number of epochs to train (default: 10).

    --lr
    Learning rate (default: 0.0004).

    --spawn-rate
    Probabilty of spawning a point (default: 1.0).

    --max-trans
    The scalar on the translation we generate and predict (default: 0.1).

    --max-spawn
    How many flurophores are spawned total (default: 1).

    --save-stats
    Save the stats of the training for later graphing.

    --no-cuda
    Disables CUDA training.

    --deterministic
    Run deterministically.

    --normalise-basic
    Normalise with torch basic intensity divide.

    --seed
    Random seed (default: 1).

    --cont
    Continuous sigma values

    --log-interval
    How many batches to wait before logging training status (default 100)

    --num-points
    How many points to optimise (default 200).

    --aug
    Do we augment the data with XY rotation (default False)?

    --num-aug
    How many augmentations to perform per datum (default 10).

    --save-interval
    How many batches to wait before saving (default 1000).

    --load
    A checkpoint file to load in order to continue training.

    --savename
    The name for checkpoint save file.

    --savedir
    The name for checkpoint save directory.

    --allocfile
    An optional data order allocation file.

    --sigma-file
    Optional file for the sigma blur dropoff.

    --dropout
    When coupled with objpath, what is the chance of a point being dropped? (default 0.0)

    --wobble
    Distance to wobble our fluorophores (default 0.0)

    --fitspath
    Path to a directory of FITS files.

    --objpath
    Path to the obj for generating data
    
    --train-size
    The size of the training set (default: 50000)

    --image-height
    The height of the images involved (default: 128).

    --image-width
    The width of the images involved (default: 128).

    --test-size
    The size of the training set (default: 200)

    --valid-size
    The size of the training set (default: 200)

    --buffer-size
    How big is the buffer in images?  (default: 40000)

## Publication

[3D Structure from 2D Microscopy images using Deep Learning - Frontiers in Bioinformatics](https://www.frontiersin.org/articles/10.3389/fbinf.2021.740342/abstract)

    @article{blundell3DStructure2D01,
        title = {{{3D Structure}} from {{2D Microscopy}} Images Using {{Deep Learning}}},
        author = {Blundell, Benjamin James and Rosten, Ed and Ch'ng, QueeLim and Cox, Susan and Manley, Suliana and Sieben, Christian},
        year = {2021},
        journal = {Frontiers in Bioinformatics},
        volume = {0},
        publisher = {{Frontiers}},
        issn = {2673-7647},
        doi = {10.3389/fbinf.2021.740342},
        copyright = {All rights reserved},
        langid = {english},
    }

### Citing this software with Bibtex

See the CITATION.cff file.

## Contributing

You can contribute to this project by submitting a pull request through github. Suggestions and feedback are welcomed through Github's Issue tracker. If you want to contact the authors please email Benjamin Blundell at benjamin.blundell@kcl.ac.uk. 

### Useful links

* Docker image based on - [https://github.com/anibali/docker-pytorch](https://github.com/anibali/docker-pytorch)

*"And the moral of the story; appreciate what you've got, because basically, I'm fantastic!" - Holly. Red Dwarf*


changelog.txt
2022-02-16  Benjamin Blundell  <benjamin.blundell@kcl.ac.uk>

    * Custom image sizes now an option.

    * Renderer now takes fewer options as perspective matrix isn't used.

    * All steps of the data pipeline are now typed.

    * Removed various options such as predict sigma (we always predict), no-translate (we always translate) and several others to streamline things.

    * Separated out train and test functions into different files to aid organisation.

    * Added the basics of detecting whether or not the model has stopped improving, using the ICP algorithm and numba.

    * Changed the tests a little to reflect the various changes elsewhere in the program.

    * Various minor improvements all over the place.


train.py
""" # noqa
   ___           __________________  ___________
  / _/__  ____  / __/ ___/  _/ __/ |/ / ___/ __/
 / _/ _ \/ __/ _\ \/ /___/ // _//    / /__/ _/      # noqa
/_/ \___/_/   /___/\___/___/___/_/|_/\___/___/      # noqa
Author : Benjamin Blundell - benjamin.blundell@kcl.ac.uk

train.py - an attempt to find the 3D shape from an image.
To train a network, use:
  python train.py <OPTIONS>

See the README file and the __main__ function for the
various options.

"""

import torch
import torch.optim as optim
import numpy as np
import math
import random
import argparse
import os
import sys
from util.points import init_points_poisson, load_points, save_points, init_points, init_points_spot
from util.loadsave import save_checkpoint, save_model
from data.loader import Loader
from data.imageload import ImageLoader
from data.sets import DataSet, SetType
from data.buffer import Buffer, BufferImage
from stats import stats as S
from net.renderer import Splat
from net.net import Net
from util.math import PointsTen
from train.train import train


def init(args, device):
    """
    Initialise all of our models, optimizers and other useful
    things before passing on to train.

    Parameters
    ----------
    args : dict
        The arguments object created in the __main__ function.
    device : str
        The device to run the model on (cuda / cpu)

    Returns
    -------
    None
    """

    # Continue training or start anew
    # Declare the variables we absolutely need
    model = None
    points = None
    buffer_train = None
    buffer_test = None
    data_loader = None
    optimiser = None

    train_set_size = args.train_size
    valid_set_size = args.valid_size
    test_set_size = args.test_size

    if args.aug:
        train_set_size = args.train_size * 4
        valid_set_size = args.valid_size * 4
        test_set_size = args.test_size * 4
    # Sigma checks. Do we use a file, do we go continuous etc?
    # Check for sigma blur file

    sigma_lookup = [10.0, 1.25]
    if len(args.sigma_file) > 0:
        if os.path.isfile(args.sigma_file):
            with open(args.sigma_file, "r") as f:
                ss = f.read()
                sigma_lookup = []
                tokens = ss.replace("\n", "").split(",")
                for token in tokens:
                    sigma_lookup.append(float(token))

    # Setup our splatting pipeline. We use two splats with the same
    # values because one never changes its points / mask so it sits on
    # the gpu whereas the dataloader splat reads in differing numbers of
    # points.

    splat_in = Splat(
        device=device,
        size=(args.image_height, args.image_width),
    )
    splat_out = Splat(
        device=device,
        size=(args.image_height, args.image_width),
    )

    # Setup the dataloader - either generated from OBJ or fits
    if args.fitspath != "":
        data_loader = ImageLoader(
            size=args.train_size + args.test_size + args.valid_size,
            image_path=args.fitspath,
            sigma=sigma_lookup[0],
        )

        set_train = DataSet(
            SetType.TRAIN, train_set_size, data_loader, alloc_csv=args.allocfile
        )
        set_test = DataSet(SetType.TEST, test_set_size, data_loader)
        set_validate = DataSet(SetType.VALID, valid_set_size, data_loader)

        buffer_train = BufferImage(
            set_train,
            buffer_size=args.buffer_size,
            device=device,
            image_size=(args.image_height, args.image_width),
        )
        buffer_test = BufferImage(
            set_test,
            buffer_size=test_set_size,
            image_size=(args.image_height, args.image_width),
            device=device,
        )
        buffer_valid = BufferImage(
            set_validate,
            buffer_size=valid_set_size,
            image_size=(args.image_height, args.image_width),
            device=device,
        )

    elif args.objpath != "":
        data_loader = Loader(
            size=args.train_size + args.test_size + args.valid_size,
            objpath=args.objpath,
            wobble=args.wobble,
            dropout=args.dropout,
            spawn=args.spawn_rate,
            max_spawn=args.max_spawn,
            translate=(not args.no_data_translate),
            sigma=sigma_lookup[0],
            max_trans=args.max_trans,
            augment=args.aug,
            num_augment=args.num_aug,
        )

        fsize = min(data_loader.size - test_set_size, train_set_size)
        set_train = DataSet(SetType.TRAIN, fsize, data_loader, alloc_csv=args.allocfile)
        set_test = DataSet(SetType.TEST, test_set_size, data_loader)
        set_validate = DataSet(SetType.VALID, valid_set_size, data_loader)

        buffer_train = Buffer(
            set_train, splat_in, buffer_size=args.buffer_size, device=device
        )

        buffer_test = Buffer(
            set_test, splat_in, buffer_size=test_set_size, device=device
        )

        buffer_valid = Buffer(
            set_validate, splat_in, buffer_size=valid_set_size, device=device
        )
    else:
        raise ValueError("You must provide either fitspath or objpath argument.")

    # TODO - possibly remove fast-forward and what not.
    # TODO - Loading for retraining should go somewhere else. We hardly ever
    # do that these days anyway

    points = init_points(
        args.num_points, device=device, deterministic=args.deterministic
    )

    if args.ipspot:
        points = init_points_spot(
            args.num_points, device=device, deterministic=args.deterministic
        )
    elif args.poisson:
        points = init_points_poisson(
            args.num_points, device=device
        )

    model = Net(
        splat_out,
        max_trans=args.max_trans,
    ).to(device)

    if args.poseonly:
        from util.plyobj import load_obj, load_ply

        if "obj" in args.objpath:
            points = load_obj(objpath=args.objpath)
        elif "ply" in args.objpath:
            points = load_ply(args.objpath)
        else:
            raise ValueError("If using poseonly, objpath must be set.")

        points = PointsTen(device=device).from_points(points)

    else:
        # Load our init points as well, if we are loading the same data
        # file later on - this is only in initialisation
        if os.path.isfile(args.savedir + "/points.csv"):
            print("Loading points file", args.savedir + "/points.csv")
            tpoints = load_points(args.savedir + "/points.csv")
            points = PointsTen(device=device)
            points.from_points(tpoints)
        else:
            save_points(args.savedir + "/points.csv", points)

        points.data.requires_grad_(requires_grad=True)

    # Save the training and test data to disk so we can interrogate it later
    set_train.save(args.savedir + "/train_set.pickle")
    set_test.save(args.savedir + "/test_set.pickle")
    data_loader.save(args.savedir + "/train_data.pickle")

    variables = []
    variables.append({"params": model.parameters(), "lr": args.lr})

    if not args.poseonly:
        variables.append({"params": points.data, "lr": args.plr})

    optimiser = optim.AdamW(variables)

    print("Starting new model")

    # Now start the training proper
    train(
        args,
        device,
        sigma_lookup,
        model,
        points,
        buffer_train,
        buffer_test,
        buffer_valid,
        data_loader,
        optimiser,
    )

    save_model(model, args.savedir + "/model.tar")


if __name__ == "__main__":
    # Training settings
    # TODO - potentially too many options now so go with a conf file?
    parser = argparse.ArgumentParser(description="PyTorch Shaper Train")
    parser.add_argument(
        "--batch-size",
        type=int,
        default=20,
        help="input batch size for training \
                          (default: 20)",
    )
    parser.add_argument(
        "--epochs",
        type=int,
        default=10,
        help="number of epochs to train (default: 10)",
    )
    parser.add_argument(
        "--lr",
        type=float,
        default=0.004,
        help="learning rate (default: 0.004)",
    )
    parser.add_argument(
        "--plr",
        type=float,
        default=0.0004,
        help="learning rate (default: 0.0004)",
    )
    parser.add_argument(
        "--spawn-rate",
        type=float,
        default=1.0,
        help="Probabilty of spawning a point \
                          (default: 1.0).",
    )
    parser.add_argument(
        "--max-trans",
        type=float,
        default=0.1,
        help="The scalar on the translation we generate and predict \
                          (default: 0.1).",
    )
    parser.add_argument(
        "--reduction",
        type=float,
        default=0.75,
        help="The factor to reduce the learning rate by \
                          (default: 0.75).",
    )
    parser.add_argument(
        "--max-spawn",
        type=int,
        default=1,
        help="How many flurophores are spawned total. \
                          (default: 1).",
    )
    parser.add_argument(
        "--save-stats",
        action="store_true",
        default=False,
        help="Save the stats of the training for later \
                          graphing.",
    )
    parser.add_argument(
        "--no-cuda", action="store_true", default=False, help="disables CUDA training"
    )
    parser.add_argument(
        "--deterministic",
        action="store_true",
        default=False,
        help="Run deterministically",
    )
    parser.add_argument(
        "--ipspot",
        action="store_true",
        default=False,
        help="Initialise points across space with a gaussian and a spot in the middle.\
                        (default: false)",
    )
    parser.add_argument(
        "--poisson",
        action="store_true",
        default=False,
        help="Initialise points across space with a gaussian and a spot in the middle.\
                        (default: false)",
    )
    parser.add_argument(
        "--no-data-translate",
        action="store_true",
        default=False,
        help="Turn off translation in the data \
                            loader(default: false)",
    )
    parser.add_argument(
        "--normalise-basic",
        action="store_true",
        default=False,
        help="Normalise with torch basic intensity divide",
    )
    parser.add_argument(
        "--seed", type=int, default=1, metavar="S", help="random seed (default: 1)"
    )
    parser.add_argument(
        "--log-interval",
        type=int,
        default=100,
        metavar="N",
        help="how many batches to wait before logging training \
                          status",
    )
    parser.add_argument(
        "--pinterval",
        type=int,
        default=1000,
        help="how many steps to wait before checking the points learning rate (default: 1000)",
    )
    parser.add_argument(
        "--num-points",
        type=int,
        default=200,
        help="how many points to optimise (default 200)",
    )
    parser.add_argument(
        "--aug",
        default=False,
        action="store_true",
        help="Do we augment the data with XY rotation (default False)?",
        required=False,
    )
    parser.add_argument(
        "--num-aug",
        type=int,
        default=10,
        help="How many augmentations to perform (default: 10)",
    )
    parser.add_argument(
        "--poseonly",
        default=False,
        action="store_true",
        help="Only optimise the pose. Default false",
        required=False,
    )
    parser.add_argument(
        "--adapt",
        default=False,
        action="store_true",
        help="Adaptive learning rate (default: False)",
        required=False,
    )
    parser.add_argument(
        "--image-width",
        type=int,
        default=320,
        help="The width of the input and output images \
                          (default: 320).",
    )
    parser.add_argument(
        "--image-height",
        type=int,
        default=150,
        help="The height of the input and output images \
                          (default: 150).",
    )
    parser.add_argument(
        "--save-interval",
        type=int,
        default=1000,
        help="how many batches to wait before saving.",
    )
    parser.add_argument(
        "--load",
        help="A checkpoint file to load in order to continue \
                          training",
    )
    parser.add_argument(
        "--savename",
        default="checkpoint.pth.tar",
        help="The name for checkpoint save file.",
    )
    parser.add_argument(
        "--savedir", default="./save", help="The name for checkpoint save directory."
    )
    parser.add_argument(
        "--allocfile", default=None, help="An optional data order allocation file."
    )
    parser.add_argument(
        "--sigma-file", default="", help="Optional file for the sigma blur dropoff."
    )
    parser.add_argument(
        "--dropout",
        type=float,
        default=0.0,
        help="When coupled with objpath, what is the chance of \
                          a point being dropped? (default 0.0)",
    )
    parser.add_argument(
        "--wobble",
        type=float,
        default=0.0,
        help="Distance to wobble our fluorophores \
                          (default 0.0)",
    )
    parser.add_argument(
        "--fitspath",
        default="",
        help="Path to a directory of FITS files.",
        required=False,
    )
    parser.add_argument(
        "--objpath",
        default="",
        help="Path to the obj for generating data",
        required=False,
    )
    parser.add_argument(
        "--train-size",
        type=int,
        default=50000,
        help="The size of the training set (default: 50000)",
    )
    parser.add_argument(
        "--test-size",
        type=int,
        default=200,
        help="The size of the training set (default: 200)",
    )
    parser.add_argument(
        "--valid-size",
        type=int,
        default=200,
        help="The size of the training set (default: 200)",
    )
    parser.add_argument(
        "--buffer-size",
        type=int,
        default=40000,
        help="How big is the buffer in images? \
                          (default: 40000)",
    )
    args = parser.parse_args()

    # Stats turn on
    if args.save_stats:
        S.on(args.savedir)

    # Initial setup of PyTorch
    use_cuda = not args.no_cuda and torch.cuda.is_available()
    torch.manual_seed(args.seed)
    device = torch.device("cuda" if use_cuda else "cpu")
    kwargs = {"num_workers": 1, "pin_memory": True} if use_cuda else {}
    print("Using device", device)

    if args.deterministic:
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False
        torch.manual_seed(args.seed)
        np.random.seed(args.seed)
        random.seed(args.seed)

    init(args, device)
    print("Finished Training")
    S.close()
